### PR TITLE
List Versions

### LINKED ISSUE #644: Populate rgw_bucket_dir_entry owner information
### What needs to be done

See SQLiteList::objects. Populate owner information in listings. Ideally only if requested (ListObjectsV2 fetch-owner header).

### Why it needs to be done

S3 compatibility

### Acceptance Criteria

Revert commit rgw/sfs: Remove checks for bucket list owner, add checks to src/test/rgw/sfs/test_rgw_sfs_sfs_bucket.cc

### LINKED ISSUE #647: rgw/sfs: Rewrite bucket _version_ listing leveraging SQL queries and indices
Follow up to the bucket listing issue #509
See that ticket for details on testing

### PR DESCRIPTION
**Pull request by irq0**
Reimplementation of list versions and integration into bucket->list().
Leverage SQL (instead of get_all) and support pagination.
Populate object owner information. 

Add unit tests for corner cases, as direct s3-tests coverage is weaker than the regular list coverage. Indirectly lifecycle tests rely heavily on list versions. 

Fixes: https://github.com/aquarist-labs/s3gw/issues/647
Fixes: https://github.com/aquarist-labs/s3gw/issues/644
Epic: https://github.com/aquarist-labs/s3gw/issues/256

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

### CODE REVIEW

File: src/rgw/driver/sfs/bucket.cc [hunk 1]
```diff
@@ -148,149 +147,96 @@ int SFSBucket::list(
...
-  if (list.objects(
+  // Version listing on unversioned buckets is equivalent to object listing
+  const bool want_list_versions =
+      versioning_enabled() ? params.list_versions : false;
+  const bool listing_succeeded =
+      (want_list_versions && list.versions(
+                                 get_bucket_id(), params.prefix, start_with,
+                                 max, results.objs, &results.is_truncated
+                             )) ||
+      list.objects(
```
Comment (jecluis) [inline src/rgw/driver/sfs/bucket.cc]: Maybe I'm missing something here, but won't this be running the `list.objects()` even if we `want_list_versions` but fail at `list.versions()` ? Is that meant by design?

File: src/rgw/driver/sfs/bucket.cc [hunk 2]
```diff
@@ -148,149 +147,96 @@ int SFSBucket::list(
...
-      }
+  if (!params.delim.empty()) {
+    std::vector<rgw_bucket_dir_entry> new_results;
+    list.roll_up_common_prefixes(
+        params.prefix, params.delim, results.objs, results.common_prefixes,
+        new_results
+    );
+
+    // Is there actually more after the rolled up common prefix? We
+    // can't tell form the original object query. Ask again for
```
Comment (jecluis) [inline src/rgw/driver/sfs/bucket.cc]: s/form/from/

File: src/rgw/driver/sfs/bucket.cc [hunk 3]
```diff
@@ -148,149 +147,96 @@ int SFSBucket::list(
...
-        const std::string& last =
-            std::prev(results.common_prefixes.end())->first;
-        results.next_marker = rgw_obj_key(last);
-      }
+  if (!params.delim.empty()) {
+    std::vector<rgw_bucket_dir_entry> new_results;
+    list.roll_up_common_prefixes(
+        params.prefix, params.delim, results.objs, results.common_prefixes,
+        new_results
+    );
```
Comment (jecluis) [inline src/rgw/driver/sfs/bucket.cc:183]: From what I can tell, won't this essentially filter the `result.objs` contents? If that's the case, isn't it possible that we will end up with less than `max` entries in the resulting `results.objs = new_results` (later in the function), even if the results have been truncated during the listing performed above?

File: src/rgw/driver/sfs/sqlite/sqlite_buckets.cc [hunk 4]
```diff
@@ -42,6 +45,22 @@ std::optional<DBOPBucketInfo> SQLiteBuckets::get_bucket(
...
+  const auto rows = storage.select(
+      columns(&DBUser::user_id, &DBUser::display_name),
+      inner_join<DBUser>(on(is_equal(&DBBucket::owner_id, &DBUser::user_id))),
+      where(is_equal(&DBBucket::bucket_id, bucket_id))
+  );
+  if (rows.size() == 0) {
+    return std::nullopt;
+  }
+  const auto row = rows[0];
+  return std::make_pair(std::get<0>(row), std::get<1>(row).value());
```
Comment (jecluis) [inline src/rgw/driver/sfs/sqlite/sqlite_buckets.cc]: feels like the either the first `get<>()` is missing a `.value()`, or the `.value()` on the second `get<>()` should not be there.

File: src/rgw/driver/sfs/sqlite/sqlite_list.cc [hunk 5]
```diff
@@ -78,7 +78,104 @@ bool SQLiteList::objects(
...
+      // first versions, then delete markers
+      // newest to oldest version
+      multi_order_by(
+          order_by(&DBObject::name).asc(),
+          order_by(&DBVersionedObject::version_type).asc(),
+          order_by(&DBVersionedObject::commit_time).desc(),
+          order_by(&DBVersionedObject::id).desc()
+      ),
+      limit(query_limit)
+  );
```
Comment (jecluis) [inline src/rgw/driver/sfs/sqlite/sqlite_list.cc:164]: okay, so this query made my ears bleed a bit, but it looks sound. I'm trusting it blindly.

**Review by jecluis — CHANGES_REQUESTED** [review]
more of a "request for clarification" than a "request changes". It looks good though.

File: src/rgw/driver/sfs/bucket.cc (hunk 3 above)
Comment (irq0) [reply src/rgw/driver/sfs/bucket.cc:183]: I'd say it "compresses" `result.objs`. Multiple objs may become one common prefix.

We return less than `max` objs if len(common prefixes) > 0. 

I could not find a definitive answer if `max` applies to just objects returned or includes common prefixes. 
However, the docs said it is OK to return less than max. Important for proper pagination is the `next_marker`. Clients must use that to iterate a bucket - there is no start after 1000 objects.

To make the queries simpler this interprets `max` as `max(objects with search criteria)` and common prefixes as just a way to compress results. If we were to maximize results we would need more complex common prefix searches to find up to max of them. I'd say this is future work if we find that it would improve performance of clients doing lots of delimiter lookups. The less we return the more requests a client has to do. 

Fortunately, there are s3 tests that exercise the pagination and common prefix bits. They helped a lot and I hope that I did not just fit their interpretation ;)

File: src/rgw/driver/sfs/bucket.cc (hunk 1 above)
Comment (irq0) [reply src/rgw/driver/sfs/bucket.cc]: Good catch. Tired to be clever..

File: src/rgw/driver/sfs/sqlite/sqlite_buckets.cc (hunk 4 above)
Comment (irq0) [reply src/rgw/driver/sfs/sqlite/sqlite_buckets.cc]: Thanks for pointing this out. It is actually neither. The second is an optional and value() should have been a value_or("")

**Review by jecluis — APPROVED** [review]
